src/objective_types.jl

```diff
@@ -3,41 +3,41 @@ real_to_complex(d::AbstractObjective, x) = iscomplex(d) ? real_to_complex(x) : x
 complex_to_real(d::AbstractObjective, x) = iscomplex(d) ? complex_to_real(x) : x
 
 # Used for objectives and solvers where no gradient is available/exists
-mutable struct NonDifferentiable{T,Tcplx} <: AbstractObjective where {T<:Real,
+mutable struct NonDifferentiable{T,A<:AbstractArray{T},Tcplx} <: AbstractObjective where {T<:Real,
                                                             Tcplx<:Union{Val{true},Val{false}}  #if true, must convert back on every f call
                                                             }
     f
     f_x::T
-    last_x_f::Array{T}
+    last_x_f::A
```

Comment:
What's the advantage of doing this, instead of `last_x_f::AbstractArray{T}` ?

Reply:
The compiler can infer the type of `last_x_f` and generate high performance code. In the documentation, at https://docs.julialang.org/en/stable/manual/performance-tips/#Avoid-fields-with-abstract-type-1 and https://docs.julialang.org/en/stable/manual/performance-tips/#Avoid-fields-with-abstract-containers-1, it is explained in more detail why one should not use fields with abstract types.

Reply:
Ah, thanks! That's probably something I've messed up in a lot of my code the last months.